src/main/java/edu/cornell/kfs/pmw/document/validation/impl/PaymentWorksVendorGlobalRule.java

```diff
@@ -0,0 +1,162 @@
+        if (ObjectUtils.isNull(pmwVendor)) {
+            putFieldError(errorPropertyPath,
+                    PaymentWorksKeyConstants.ERROR_PAYMENTWORKS_VENDOR_GLOBAL_DETAILS_ADDLINE_NOT_FOUND,
+                    String.valueOf(vendorDetail.getPmwVendorId()));
+            success = false;
+        }
+
+        return success;
+    }
+
+    private String buildVendorDetailsAddLinePropertyPath(final String propertyName) {
+        return StringUtils.join(KFSConstants.ADD_PREFIX, KFSConstants.DELIMITER,
+                PaymentWorksPropertiesConstants.VENDOR_DETAILS, KFSConstants.DELIMITER, propertyName);
+    }
+
+    @Override
+    protected boolean processCustomRouteDocumentBusinessRules(final MaintenanceDocument document) {
+        setupConvenienceObjects();
+        boolean success = super.processCustomRouteDocumentBusinessRules(document);
+        success &= checkActionType();
+        success &= checkPaymentWorksVendorChanges();
+        return success;
+    }
+
+    @Override
+    protected boolean processCustomApproveDocumentBusinessRules(final MaintenanceDocument document) {
+        setupConvenienceObjects();
+        boolean success = super.processCustomApproveDocumentBusinessRules(document);
+        success &= checkActionType();
+        success &= checkPaymentWorksVendorChanges();
+        return success;
+    }
+
+    private boolean checkActionType() {
+        PaymentWorksVendorGlobalAction actionType;
+        try {
+            actionType = paymentWorksVendorGlobal.getActionType();
+        } catch (RuntimeException e) {
+            actionType = null;
+        }
+        if (actionType == null || actionType != PaymentWorksVendorGlobalAction.RESTAGE_FOR_UPLOAD) {
+            putFieldError(PaymentWorksPropertiesConstants.ACTION_TYPE_CODE,
+                    PaymentWorksKeyConstants.ERROR_PAYMENTWORKS_VENDOR_GLOBAL_ACTION_INVALID,
+                    PaymentWorksVendorGlobalAction.RESTAGE_FOR_UPLOAD.actionLabel);
+            return false;
+        }
+        return true;
+    }
+
+    private boolean checkPaymentWorksVendorChanges() {
+        final List<PaymentWorksVendorGlobalDetail> vendorDetails = paymentWorksVendorGlobal.getVendorDetails();
+        if (CollectionUtils.isEmpty(vendorDetails)) {
+            putFieldError(PaymentWorksPropertiesConstants.VENDOR_DETAILS,
+                    PaymentWorksKeyConstants.ERROR_PAYMENTWORKS_VENDOR_GLOBAL_DETAILS_EMPTY);
+            return false;
+        }
+
+        boolean success = true;
+        int i = -1;
+        final Set<Integer> encounteredPmwVendors = new HashSet<>();
+
+        for (final PaymentWorksVendorGlobalDetail vendorDetail : vendorDetails) {
+            i++;
+            final String errorPathPrefix = buildVendorDetailsErrorPathPrefix(i);
+            final String pmwVendorIdErrorPath = errorPathPrefix + PaymentWorksPropertiesConstants.PMW_VENDOR_ID;
+            if (!encounteredPmwVendors.add(vendorDetail.getPmwVendorId())) {
```

Comment:
is this if statement correct?
Shouldn't it be a check to see if the vendor is already in the encounteredPmwVendors?

Reply:
`encounteredPmwVendors` is a `Set`, so its `add()` method will return true if it did NOT already contain the given element or false otherwise. Since the body of the block is only meant to execute when a duplicate is encountered, I believe the statement's logic is correct.

Reply:
Maybe this could be refactored so that is clearer?

Reply:
I'll add a local boolean variable to clarify the intent.